providers/implementations/kdfs/hkdf.c

```diff
@@ -427,6 +441,83 @@ const OSSL_DISPATCH ossl_kdf_hkdf_functions[] = {
     OSSL_DISPATCH_END
 };
 
+static void *kdf_hkdf_fixed_digest_new(void *provctx, const char *digest)
+{
+    OSSL_LIB_CTX *libctx = PROV_LIBCTX_OF(provctx);
+    KDF_HKDF *ctx;
+    OSSL_PARAM params[2];
+
+    ctx = kdf_hkdf_new(provctx);
+    if (ctx == NULL)
+        return NULL;
+
+    params[0] = OSSL_PARAM_construct_utf8_string(OSSL_ALG_PARAM_DIGEST,
+                                                 (char *)digest, 0);
+    params[1] = OSSL_PARAM_construct_end();
+    if (!ossl_prov_digest_load_from_params(&ctx->digest, params, libctx)) {
+        kdf_hkdf_free(ctx);
+        return NULL;
+    }
+
+    return ctx;
+}
+
+#define KDF_HKDF_FIXED_DIGEST_NEW(hashname, hashstring) \
+    static void *kdf_hkdf_##hashname##_new(void *provctx) \
+    { \
+        return kdf_hkdf_fixed_digest_new(provctx, hashstring); \
+    }
+
+KDF_HKDF_FIXED_DIGEST_NEW(sha256, "SHA256")
+KDF_HKDF_FIXED_DIGEST_NEW(sha384, "SHA384")
+KDF_HKDF_FIXED_DIGEST_NEW(sha512, "SHA512")
+
+static const OSSL_PARAM *kdf_hkdf_fixed_digest_settable_ctx_params(ossl_unused void *ctx,
+                                                                   ossl_unused void *provctx)
+{
+    static const OSSL_PARAM known_settable_ctx_params[] = {
+        HKDF_COMMON_SETTABLES_NO_DIGEST,
+        OSSL_PARAM_octet_string(OSSL_KDF_PARAM_INFO, NULL, 0),
+        OSSL_FIPS_IND_SETTABLE_CTX_PARAM(OSSL_KDF_PARAM_FIPS_KEY_CHECK)
+        OSSL_PARAM_END
+    };
+    return known_settable_ctx_params;
+}
+
+static int kdf_hkdf_fixed_digest_set_ctx_params(void *vctx, const OSSL_PARAM params[])
+{
+    if (OSSL_PARAM_locate_const(params, OSSL_ALG_PARAM_DIGEST) != NULL) {
+        /*
+         * TODO: OPENSSL MAINTAINERS: should it fail an any attempt to set the digest or
+         * should it be allowed to set the digest to the current digest?
+         */
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_DIGEST_NOT_ALLOWED,
```

Comment:
If the user tries to set the digest on a fixed-digest HKDF, should it always fail?  Only fail if the new digest is different?

Reply:
Interesting question.  I don't think failing when given the `OSSL_ALG_PARAM_DIGEST` parameter after advertising (via `kdf_hkdf_fixed_digest_settable_ctx_params`) that it isn't accepted is correct.  Parameter handling is deliberately designed to ignore unknown parameters and by not including it in the settable list, it is unknown.

Supporting it and failing if it is changed seems like reasonable user-friendly behaviour.  I'd suggest not separating the settable parameter handling and instead have a flag or fixed digest field in the `KDF_HKDF` structure.

Reply:
I will make this more user-friendly change.

Clarification: If the user tries to set the HKDF's digest to a different digest value it will fail.  If they try to set it to the fixed digest value it will succeed.